### ISSUE TITLE
Confidence on Estimates

### ISSUE THREAD

**Issue by gotdibbs** [Zuehlke/poinz#178]
Hey there!

I was thinking it could be nice to add some way to give a confidence level when you vote for an estimate. So like some way to visually show "I'm guessing this is a `3` but I've never done this before" vs. "We've done this 100 times before, it is a `1`".

I think that could help with facilitating discussions, but I'm not sure exactly how much.

Thoughts? Thanks!

**Comment by xeronimus** [comment]
Hey @gotdibbs ,  thanks for your suggestion!

Am I right in assuming that this "confidence level" would be visible only after a story is revealed? 
As always, we need to discuss the "disadvantage of added complexity" versus the benefit of the new feature.


- what is the added benefit:  in large groups, the discussion is efficient?  
- are there downsides?  e.g.  "anchoring" effect -> might influence others heavily.  
- what is the simplest form of this feature?  user can choose one of 3 levels :  very unsure < normal < very confident.  where "normal" would be the default.

**Comment by gotdibbs** [comment]
You bet!

> Am I right in assuming that this "confidence level" would be visible only after a story is revealed?
As always, we need to discuss the "disadvantage of added complexity" versus the benefit of the new feature.

I think so, yeah. Basically it is information given at the time of estimation, and then leveraged as part of the discussion.

> What is the added benefit: in large groups the discussion is efficient?

I think I see the value as clearly seeing that all of the people who pointed something high, were unsure, and thus we need to have more discussion. Ideally they'd just do a `?` but there can be pressure, especially in newer teams, to make a guess of something and just assume its complex if they don't know anything about it.

> Are there downsides?

I could see this being potentially cumbersome in the UI, so it might be nice if you could toggle it on or off as a feature. As far as downsides for the actual process, I think it could lead to some bikeshedding potentially, but having another data point I don't see as hurting ... just helping more quiet folks to better express their opinions.

> What is the simplest form of this feature?

I see this as confident vs. unconfident honestly, but an normal or "undefined" state could be ok too ... thought I'm not sure that middle state needs UI attached to it. You could say click an up arrow to denote high confidence, and click it again to unset your confidence level.

**Comment by xeronimus** [comment]
## Implementation details

### New flag on room: confidenceLevels

by default, the "confidence level" feature would be disabled. the UI would not display the option to set the confidence level when a user estimates. the UI would not display confidence levels when a story is revealed.

This would mean a new command "toggleConfidenceLevels" and two new events "confidenceLevelsOn" and "confidenceLevelsOff"  (similar to [toggleAutoReveal](https://github.com/Zuehlke/poinz/blob/master/docu/commandAndEventDocu.md#command-toggleautoreveal-))

and of course the "settings" UI in order to toggle it.


### confidenceLevel in giveStoryEstimate

add a new payload attribute to [giveStoryEstimate ](https://github.com/Zuehlke/poinz/blob/master/docu/commandAndEventDocu.md#giveStoryEstimate) command:  "confidenceLevel".  confidenceLevel is an optional attribute. can have values 
* -1 =  I'm unsure / I'm guessing.  
* 0 = default / not set / "normal confidence"
* 1 = I'm very confident

### confidenceLevel stored in room

currently the estimations are stored like this:

```json
{
    "stories": [
        {
            "id": "the-id-of-the-story",
            "title": "some story",
            "estimations": {
                "user-id-1": 5,
                "user-id-2": 8
            }
        }
    ]
}
```

we would need to add the "confidenceLevel" per estimation. (0 by default / if not set)

```json
{
    "stories": [
        {
            "id": "the-id-of-the-story",
            "title": "some story",
            "estimations": {
                "user-id-1": {
                    "value": 5,
                    "confidenceLevel": 0
                },
                "user-id-2": {
                    "value": 8,
                    "confidenceLevel": -1
                }
            }
        }
    ]
}
```

## UI

let's think about how the UI would look like....

**Comment by gvheertum** [comment]
Regarding the UI, you could consider having the confidence picked in a selection before the points them selves. By default you could set it to "neutral" and allow the user to switch to uncertain or very certain.

In the points breakdown you could then perhaps consider exclusion/spotlight of these values (e.g. exclude uncertain). I think this could potentially be very helpful for mixed skill teams (or teams with newcomers).

**Comment by xeronimus** [comment]
I implemented a working "confidence" feature. I'd be glad for some feedback:

## Room Setting
Users can now enable confidence levels in the room settings. This option is disabled by default (when creating a new room).

![poinz_conflevels_room_settings](https://user-images.githubusercontent.com/1777143/134805198-3ed50388-81fe-4dba-a46b-f584541e30fb.png)

What do you think about the wording?

## Specifying own confidence when estimating
If the confidence level feature is toggled for this room, users can specify their confidence when estimating a story.

![poinz_conflevels_specifying_conflevel](https://user-images.githubusercontent.com/1777143/134805216-b6322b8e-89c5-4f90-a067-0d827d8b7ff3.png)

This is always set to "Default" when estimating a new story.

## Estimation Summary / Estimations reveal
After the estimations are revealed, cards are marked if the user specified either unsure (question mark) or very sure (exclamation mark).

In this example, Sergio was not sure at all, chromie left the setting "default", Foxy is very sure about this story being 21.
![poinz_conflevels_estm_summary_A](https://user-images.githubusercontent.com/1777143/134805269-8b3ba7e6-60e9-45f4-8e7d-3b8ea514da18.png)

In the lower part of the board, the estimation summary shows all cards by default and calculates the average including all estimations.

if you toggle the checkbox, all unsure (low confidence) votes are excluded.

In this example, this leads to a higher average (altough same recommendation in this case...)

![poinz_conflevels_estm_summary_B](https://user-images.githubusercontent.com/1777143/134805304-d7124725-def5-4a61-9380-2f907f24be3a.png)

**Comment by xeronimus** [comment]
@gotdibbs @gvheertum

**Comment by gvheertum** [comment]
Looking good @xeronimus , I'll give it a try with my team this week. Will get back to you with findings. However, this looks great!

**Comment by xeronimus** [comment]
deployed 0.14.0.

see  https://github.com/Zuehlke/poinz/blob/master/docu/manual.md#estimating-with-confidence-levels

**Comment by gotdibbs** [comment]
Hey sorry! Been a crazy couple of weeks for me. What you implemented here looks great to me and makes a lot of sense. Looking forward to trying it out!
